lib/libexfat.c

```diff
@@ -451,26 +475,19 @@ int exfat_set_volume_label(struct exfat_blk_dev *bd,
 		return -1;
 	}
 
-	vol.type = EXFAT_VOLUME;
-	memset(vol.vol_label, 0, sizeof(vol.vol_label));
-	memcpy(vol.vol_label, volume_label, volume_label_len);
-	vol.vol_char_cnt = volume_label_len/2;
+	memset(pvol->vol_label, 0, sizeof(pvol->vol_label));
+	memcpy(pvol->vol_label, volume_label, volume_label_len);
+	pvol->vol_char_cnt = volume_label_len/2;
 
-	nbytes = exfat_write(bd->dev_fd, &vol, sizeof(struct exfat_dentry),
-			root_clu_off);
-	if (nbytes != sizeof(struct exfat_dentry)) {
-		exfat_err("volume entry write failed: %d\n", errno);
-		return -1;
-	}
+	loc.parent = exfat->root;
+	loc.file_offset = filter.out.file_offset;
+	loc.dev_offset = filter.out.dev_offset;
+	err = exfat_add_dentry_set(exfat, &loc, pvol, dcount, false);
+	exfat_info("new label: %s\n", label_input);
 
-	if (fsync(bd->dev_fd) == -1) {
-		exfat_err("failed to sync volume entry: %d, %s\n", errno,
-			  strerror(errno));
-		return -1;
-	}
+	free(pvol);
```

Comment:
And we have to free filter.out.dentry_set, which are allocated by exfat_lookup_dentry_set.

Reply:
See:
	err = exfat_lookup_dentry_set(exfat, exfat->root, &filter);
	if (!err) {
		pvol = filter.out.dentry_set;

even though exfat_lookup_dentry_set return error, need to free it ?